lib/vector-core/src/fanout.rs

```diff
@@ -54,174 +54,375 @@ impl Fanout {
-        while let Poll::Ready(Some(message)) = self.control_channel.poll_recv(cx) {
-            match message {
-                ControlMessage::Add(id, sink) => self.add(id, sink),
-                ControlMessage::Remove(id) => self.remove(&id),
-                ControlMessage::Replace(id, sink) => self.replace(&id, sink),
+    /// If any sink is awaiting replacement (i.e. it was temporarily replaced with `None`), read
+    /// and process messages from the control channel until that is no longer true.
+    async fn wait_for_replacements(&mut self) {
+        while self.sinks.iter().any(|x| x.1.is_none()) {
+            if let Some(msg) = self.control_channel.recv().await {
+                self.apply_control_message(msg);
+            } else {
+                // If the control channel is closed, there's nothing else we can do.
             }
         }
     }
 
-    #[inline]
-    fn handle_sink_error(&mut self, index: usize) -> Result<(), ()> {
-        // If there's only one sink, propagate the error to the source ASAP
-        // so it stops reading from its input. If there are multiple sinks,
-        // keep pushing to the non-errored ones (while the errored sink
-        // triggers a more graceful shutdown).
-        if self.sinks.len() == 1 {
-            Err(())
-        } else {
-            self.sinks.remove(index);
-            Ok(())
+    pub async fn send_stream(&mut self, events: impl Stream<Item = Event>) {
+        let stream = events.ready_chunks(1024);
+        tokio::pin!(stream);
+        while let Some(events) = stream.next().await {
+            self.send_all(events).await;
         }
     }
 
-    fn poll_sinks<F>(
-        mut self: Pin<&mut Self>,
-        cx: &mut Context<'_>,
-        poll: F,
-    ) -> Poll<Result<(), ()>>
-    where
-        F: Fn(
-            Pin<&mut (dyn Sink<Event, Error = ()> + Send)>,
-            &mut Context<'_>,
-        ) -> Poll<Result<(), ()>>,
-    {
-        self.process_control_messages(cx);
-
-        let mut poll_result = Poll::Ready(Ok(()));
-
-        let mut i = 0;
-        while let Some((_, sink)) = self.sinks.get_mut(i) {
-            if let Some(sink) = sink {
-                match poll(sink.as_mut(), cx) {
-                    Poll::Pending => poll_result = Poll::Pending,
-                    Poll::Ready(Ok(())) => (),
-                    Poll::Ready(Err(())) => {
-                        self.handle_sink_error(i)?;
-                        continue;
+    /// Send a batch of events to all connected sinks.
+    ///
+    /// This will block on the resolution of any pending reload before proceeding with the send
+    /// operation.
+    ///
+    /// # Panics
+    ///
+    /// This method can panic if the fanout receives a control message that violates some invariant
+    /// about its current state (e.g. remove a non-existent sink, etc). This would imply a bug in
+    /// Vector's config reloading logic.
+    pub async fn send_all(&mut self, events: Vec<Event>) {
```

Comment:
This will, of course, need to be adapted to `EventArray` since that is now what is being handled.